crates/wasmparser/src/validator/operators.rs

```diff
@@ -3335,6 +3336,24 @@ where
         self.pop_operand(Some(ValType::I32))?;
         Ok(())
     }
+    fn visit_i31_new(&mut self) -> Self::Output {
+        self.pop_operand(Some(ValType::I32))?;
+        self.push_operand(ValType::Ref(RefType::REF_I31))
+    }
+    fn visit_i31_get_s(&mut self) -> Self::Output {
+        match self.pop_ref()? {
+            Some(RefType::REF_I31) | Some(RefType::I31REF) => {}
+            _ => bail!(self.offset, "type mismatch: expected (ref null? i31)"),
+        };
+        self.push_operand(ValType::I32)
+    }
+    fn visit_i31_get_u(&mut self) -> Self::Output {
+        match self.pop_ref()? {
+            Some(RefType::REF_I31) | Some(RefType::I31REF) => {}
```

Comment:
Could this perhaps be modeled as `ty.heap_type() == HeapType::I31`? That feels a little more clear to me that it's accepting both nullable and non-nullable references.

Reply:
Just thought I'd follow the historical convention with those shorthand types :)

Reply:
Oh does this exist elsewhere already? If so we should probably go back and update those too (no need to do that in this PR though).